python/ml4ir/base/pipeline.py

```diff
@@ -265,9 +318,128 @@ def get_relevance_model(self, feature_layer_keys_to_fns={}) -> RelevanceModel:
+        Aggregate results of the k-fold runs and perform t-test on the results between old(prod model) and
+        new model's w.r.t the specified metrics.
+
+        Parameters
+        ----------
+        base_logs_dir : int
+            Total number of folds
+        run_id : int
+            current fold number
+        num_folds: int
+            Total number of folds
+        metrics: set()
+            Set of metrics to include in the kfold analysis
+        """
+        if not metrics:
+            return
+        rows = []
+        pvalue_threshold = 0.1
+        for i in range(num_folds):
+            row = {"fold": i}
+            logs_dir = pathlib.Path(base_logs_dir) / run_id / "fold_{}".format(i) / run_id
+            with open(logs_dir / "_SUCCESS", 'r') as f:
+                for line in f:
+                    parts = line.split(',')
+                    if parts[0] in metrics:
+                        row[parts[0]] = float(parts[1])
+            rows.append(row)
+        results = pd.DataFrame(rows)
+        self.logger.info(results)
+        # calculate statistical paired t-test
+        for dataset in ['train', 'val', 'test']:
+            t_test_stat, pvalue = stats.ttest_rel(results["{}_old_MRR".format(dataset)],
+                                                  results["{}_new_MRR".format(dataset)])
+            self.logger.info("{}_t_test_stat={}".format(dataset, t_test_stat))
+            self.logger.info(
+                "{}_pvalue={} --> (statistically significant={})".format(dataset, pvalue, pvalue < pvalue_threshold))
+
     def run(self):
         """
-        Run the pipeline to train, evaluate and save the model
+        Run the pipeline to train, evaluate and save the model. It also runs the pipeline in kfold cross validation
+        mode if specified.
+
+        Returns
+        -------
+        dict
+            Experiment tracking dictionary with metrics and metadata for the run.
+            Used for model selection and hyperparameter optimization
+
+        Notes
+        -----
+        Also populates a experiment tracking dictionary containing
+        the metadata, model architecture and metrics generated by the model
+        """
+        if self.args.kfold <= 1:
+            # Run ml4ir without kfold cross validation
+            return self.run_pipeline()
+
+        if self.args.include_testset_in_kfold:
+            if self.args.kfold < 3:
+                raise Exception("Number of folds must be > 2")
+        else:
+            if self.args.kfold < 2:
+                raise Exception("Number of folds must be > 1")
+
+        args = copy.deepcopy(self.args)
+
+        # reading, parsing the dataset (train, validation, test)
+        self.logger.info("Reading datasets ...")
+        relevance_dataset = self.get_relevance_dataset()
+        self.logger.info("Relevance Dataset created")
+
+        all_data = relevance_dataset.merge_datasets(include_testset_in_merge=args.include_testset_in_kfold)
```

Comment:
Why use args here and self.args right below?

Reply:
for that line it doesn't matter, but I agree I will change it for consistency.